mne/epochs.py

```diff
@@ -1353,6 +1361,16 @@ def _get_data(self, out=True, picks=None, item=None, verbose=None):
             data = np.empty((0, len(self.info['ch_names']), len(self.times)))
             logger.info('Loading data for %s events and %s original time '
                         'points ...' % (n_events, len(self._raw_times)))
+
+        # don't rescale if the baseline is invalid (due to cropping of the
+        # baseline period)
+        rescale_data = True
+        if self.baseline != _check_baseline(self.baseline, tmin=self.tmin,
+                                            tmax=self.tmax,
+                                            sfreq=self.info['sfreq'],
+                                            on_baseline_outside_data='adjust'):
+            rescale_data = False
```

Comment:
Relying on exact float equality is dangerous. One time when `tmin` and `sfreq` are double floating point you could get one result, and once a save-load round trip happens and they are single floating point you might get another. Is there some way instead to actually keep track of when we actually need to apply a baseline?

When I originally went to work on this at one point, I considered adding a private `self._do_baseline = False` that was `True` only in the one case when it actually needed to be applied later rather than right away, which I think is only for non-preloaded Epochs. This way you can either 1) check that `not self._do_baseline` if a `crop` operation is requested that would remove the baseline period, or (better but harder) 2) we would make the raw data getter get data from the interval `min(bmin, tmin), max(bmax, tmax)`, then crop.

Reply:
... if it seems very difficult to refactor using some `_do_baseline`, another alternative would be to have a `_baseline_isclose(a, b, sfreq)` that checked that the terms were close to within `1e-2 / sfreq` or so for example. This should be good enough to deal with floating point inaccuracies (time points should step by `1. / sfreq`, so doing `1e-2 / sfreq` allows for a lot of slop, which could be necessary for very long / hundreds-of-thousands-of-time-points epochs).

Reply:
> When I originally went to work on this at one point, I considered adding a private `self._do_baseline = False` that was `True` only in the one case when it actually needed to be applied later rather than right away, which I think is only for non-preloaded Epochs.

I have implemented this now in eca094a. And I have to admit I'm starting to get (have gotten?) lost a bit regarding all the side-effects this open-heart surgery might have. Fortunately we've got excellent test coverage, but I'm quite concerned I might be introducing bugs here. I think we should also manually test before finally merging this…

> This way you can either 1) check that `not self._do_baseline` if a `crop` operation is requested that would remove the baseline period, 

This is not necessary, as cropping only works on preloaded data, and loading applies baseline correction.

> or (better but harder) 2) we would make the raw data getter get data from the interval `min(bmin, tmin), max(bmax, tmax)`, then crop.

I haven't looked into this, I don't think I want to have this kind of complexity for now :)